rust/src/lib.rs

```diff
@@ -355,17 +387,58 @@ pub unsafe extern "C" fn zcashlc_create_account(
                 }
             })?;
 
-        db_data
+        let (account_id, usk) = db_data
             .create_account(&seed, birthday)
-            .map(|(account, usk)| {
-                let encoded = usk.to_bytes(Era::Orchard);
-                Box::into_raw(Box::new(FFIBinaryKey::new(account, encoded)))
-            })
-            .map_err(|e| anyhow!("Error while initializing accounts: {}", e))
+            .map_err(|e| anyhow!("Error while initializing accounts: {}", e))?;
+
+        let account = db_data.get_account(account_id)?.expect("just created");
+        let account_index = match account.kind() {
+            AccountKind::Derived { account_index, .. } => account_index,
+            AccountKind::Imported => unreachable!("just created"),
+        };
+
+        let encoded = usk.to_bytes(Era::Orchard);
+        Ok(Box::into_raw(Box::new(FFIBinaryKey::new(
+            account_index,
+            encoded,
+        ))))
     });
     unwrap_exc_or_null(res)
 }
 
+/// Verifies that the given seed corresponds to the viewing key for every account in the
+/// wallet.
+///
+/// Returns:
+/// - `Ok(true)` if the viewing key for every account can be derived from the provided
+///   seed.
+/// - `Ok(false)` if the derived viewing key does not match for at least one account.
+/// - `Err(_)` if a Unified Spending Key cannot be derived from the seed for an account,
+///   or an account has no known ZIP-32 derivation.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_validate_seed(
+    db_data: *const u8,
+    db_data_len: usize,
+    seed: *const u8,
+    seed_len: usize,
+    network_id: u32,
+) -> bool {
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let db_data = unsafe { wallet_db(db_data, db_data_len, network)? };
+        let seed = Secret::new((unsafe { slice::from_raw_parts(seed, seed_len) }).to_vec());
+
+        for account_id in db_data.get_account_ids()? {
+            if !db_data.validate_seed(account_id, &seed)? {
+                return Ok(false);
+            }
```

Comment:
```suggestion
            if db_data.validate_seed(account_id, &seed)? {
                return Ok(true);
            }
```
Bug: this method will currently return `true` if the database contains no accounts.

Reply:
Also, your suggestion has a bug: it early-returns, so it only checks that the _first_ account is valid with the seed, and not that _all_ accounts are valid with the seed.

Reply:
That's the intention; it's future-proof (because for imported UFVKs we won't want to validate against them.)

Reply:
Okay, I think this indicates that we do not have the same idea about what this API is meant to validate.

My understanding is that at the mobile SDK level, we currently only support a single seed phrase, and only derived accounts. Therefore, `validateSeed` checks that the database only contains accounts corresponding to this single seed phrase. If it contains _any_ accounts for a different seed phrase, or an unknown seed phrase, then we cannot use the database. _If_ we move to supporting imported UFVKs, then at that time we'd weaken this to ignore imported accounts, but we would still check all derived accounts (and from that perspective, the `WalletRead::validate_seed` API is insufficient: it combines "account is imported" and "account is derived but not from this seed" as a single `Ok(false)`, whereas they need to be distinguishable in this case).

It seems that you instead want an API that checks if the given seed phrase is _relevant_ to the database, i.e. if _any_ account can be used with that seed phrase. That might be a separately useful API, but it doesn't solve the problem of "can I use this database backup", because if the caller does not know every seed phrase for every derived account, then the answer is no.

Reply:
> It seems that you instead want an API that checks if the given seed phrase is relevant to the database, i.e. if any account can be used with that seed phrase. That might be a separately useful API, but it doesn't solve the problem of "can I use this database backup", because if the caller does not know every seed phrase for every derived account, then the answer is no.

That's not necessary for the seed phrase to be useful. Remember that when we go to spend, we take the USK as input, derive the UFVK, and then use _that_ to determine what funds we have available to spend. So we can have a database that contains UFVKs that don't correspond to the seed, and that's just fine; we won't be able to spend from those accounts.